pkg/services/dashboards/service/dashboard_service.go

```diff
@@ -82,11 +94,283 @@ type DashboardServiceImpl struct {
 	k8sclient              client.K8sHandler
 	metrics                *dashboardsMetrics
 	publicDashboardService publicdashboards.ServiceWrapper
+	serverLockService      *serverlock.ServerLockService
+	kvstore                kvstore.KVStore
+
+	dashboardPermissionsReady      chan struct{}
+	k8sDeletedDashboardsCleanupJob *k8sCleanupJob
+}
+
+type k8sCleanupJob struct {
+	stop chan struct{}
+}
+
+func (b *k8sCleanupJob) Stop() {
+	if b.stop != nil {
+		close(b.stop)
+	}
+}
+
+func (dr *DashboardServiceImpl) startK8sDeletedDashboardsCleanupJob() {
+	dr.k8sDeletedDashboardsCleanupJob = &k8sCleanupJob{
+		stop: make(chan struct{}),
+	}
+
+	go func() {
+		ticker := time.NewTicker(k8sDashboardCleanupInterval)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-dr.k8sDeletedDashboardsCleanupJob.stop:
+				return
+			case <-ticker.C:
+				if err := dr.executeCleanupWithLock(context.Background()); err != nil {
+					dr.log.Error("Failed to execute k8s dashboard cleanup", "error", err)
+				}
+			}
+		}
+	}()
+}
+
+func (dr *DashboardServiceImpl) executeCleanupWithLock(ctx context.Context) error {
+	// We're taking a leader-like locking approach here. By locking and executing, but never releasing the lock,
+	// we ensure that other instances of this service can't run in parallel and hence the cleanup will only happen once
+	// per k8sDashboardCleanupInterval by setting the maxInterval and having the time between executions be k8sDashboardCleanupInterval as well.
```

Comment:
IIUC we need to ensure that the timeout is less than the interval, so I wonder if we could just compute it on the fly to prevent misconfiguration? e.g. `timeout = interval - 1 * time.Second`?

Reply:
Sounds good